### PR TITLE
Restore building with GHC 7.10.3 by adding semigroups dependency

### PR DESCRIPTION
**Pull request by andreasabel**
Restore building with GHC 7.10.3 by simply adding `semigroups` dependency.

This also adds a CI generated automatically by `haskell-ci` in `.github/worksflows/haskell-ci.yml`.  Successful run at: https://github.com/andreasabel/semirings/actions/runs/2139986631
This CI disables the `haddock` step for GHC <= 8.4.4 because the haddock parser has some problems with `-- | @since` in `deriving` tuples.
If you do not like it (or have something better) you may delete this CI.

I have not invested restoring GHC 7.8 or below, usually this is a bit more work.

### CODE REVIEW

**Comment by Bodigrim** [comment]
@andreasabel dropping support of GHC < 8 was a conscious decision to reduce complexity we maintain, in line with other libraries, even very conservative ones. There are plenty of older releases of `semirings`, which supports GHCs back to 7.4.

**Comment by andreasabel** [comment]
> @andreasabel dropping support of GHC < 8 was a conscious decision to reduce complexity we maintain, in line with other libraries, even very conservative ones. 

Ok, maybe it was a conscious decision.  But as I prove here, it didn't reduce complexity (unless you consider one conditional standard dependency _complexity_).

> There are plenty of older releases of `semirings`, which supports GHCs back to 7.4.

We all know that using older releases causes other pains, e.g. by not being supported by stackage nightly etc.

**Comment by Bodigrim** [comment]
Yes, I consider adding a conditional block to cabal file a complexity. It also means one more configuration to test and think about. I don't want to spend precious time of my life supporting long obsolete platforms. `bytestring` and `text` do not support GHC 7.10 any longer, so I see very little sense for `semirings` to be more Catholic than the Pope.

FWIW the latest released version of `semiring` still supports GHC 7.10.

> We all know that using older releases causes other pains, e.g. by not being supported by stackage nightly etc.

Certainly Stackage Nightly does not use GHC 7.10.

**Comment by andreasabel** [comment]
In my experience, GHC 7.8 has some maintenance burden over GHC 8.0 _if you write new code_, but GHC 7.10 has very little of that (except if you need to disable warnings and use the old `-fno-warn-...` syntax),
CI, if generated automatically, also isn't a burden once it is configure.

So, I concede that making new contributions (code, CI) work for older GHCs is extra burden, but taking away working support for older GHCs is usually just reducing _perceived_ complexity.
In my projects, when I drop an older GHC, usually there is a substantial reason for this.

In general, supporting more GHCs is providing more fall-back options.  GHC bugs do not develop monotonically.  There might be undiscovered bugs in newer versions, and having more fall-back options to older GHCs may help circumnavigating a certain bug.

> FWIW the latest released version of semiring still supports GHC 7.10.

Good. But I am worried about the trend, see
- #80 

that drops GHC versions for non-substantial reasons, like haddock-incompatibility.

File: cabal.haskell-ci [hunk 1]
```diff
@@ -5,3 +5,5 @@ unconstrained:          False
 install-dependencies:   False
 copy-fields:            all
 installed:              +all -transformers
+haddock:                >= 8.6
+  -- haddock up to 8.4 cannot parse `-- | @since ...` in `deriving` tuple
```
Comment (chessai) [inline cabal.haskell-ci:9]: oh, I didn't know about this constraint. very useful.

**Comment by chessai** [comment]
We probably won't move forward with this

**Comment by andreasabel** [comment]
Fair game.
A while ago I started dropping support for GHC 7 everywhere, so this PR is obsolete now (even for me).

**Pull request closed without merging**